### PR TITLE
feat(STONEINTG-540): manual re-run integration tests

### PR DESCRIPTION
**Pull request by MartinBasti**
This PR allows to trigger manual rerun of an integration test via snapshot annotation.

## Maintainers will complete the following section

- [ ] Commit messages are descriptive enough ([hints](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/))
- [ ] Code coverage from testing does not decrease and new code is covered ([check the PR coverage on codecov](https://app.codecov.io/gh/redhat-appstudio/integration-service/pulls))
- [ ] [Controllers diagrams](https://github.com/redhat-appstudio/integration-service/tree/main/docs) are updated when PR changes controllers code  (if applicable)

### CODE REVIEW

File: gitops/snapshot.go [hunk 1]
```diff
@@ -658,3 +683,46 @@ func GetComponentSourceFromComponent(component *applicationapiv1alpha1.Component
...
 }
+
+// GetIntegrationTestRunLabelValue returns value of the label responsible for re-running tests
+func GetIntegrationTestRunLabelValue(snapshot applicationapiv1alpha1.Snapshot) (string, bool) {
+	labels := snapshot.GetLabels()
+	labelVal, ok := labels[SnaphostIntegrationTestRun]
+	return labelVal, ok
+}
+
+// TODO: take this function from operator-toolkit
```
Comment (MartinBasti) [inline gitops/snapshot.go:694]: :eyes:

**Comment by MartinBasti** [comment]
Back to draft: re-running integration tests of finished snapshot in ephemeral environments are causing deadlock, where no test is actually executed after deploying env and SEB is not removed.

File: controllers/snapshot/snapshot_adapter.go [hunk 2]
```diff
@@ -75,6 +76,87 @@ func scenariosNamesToList(integrationTestScenarions *[]v1beta1.IntegrationTestSc
...
+					return controller.RequeueWithError(err)
+				}
+				return controller.StopProcessing()
+			}
+			return controller.RequeueWithError(fmt.Errorf("failed to ensure existence of the environment for scenario %s: %w", integrationTestScenario.Name, err))
+		}
+
+	} else {
+		integrationTestScenarioStatus, ok := testStatuses.GetScenarioStatus(integrationTestScenario.Name)
+		if ok && integrationTestScenarioStatus.TestPipelineRunName != "" {
```
Comment (dirgim) [inline controllers/snapshot/snapshot_adapter.go:133]: Do we need this check here since the status for that scenario will be reset `testStatuses.ResetStatus(scenarioName)` at this point?

Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go:133]: yes, in case that for unknown reason, label failed to be removed and reconciliation may re-trigger another test.

Comment (dirgim) [reply controllers/snapshot/snapshot_adapter.go:133]: I see your point, that's certainly possible, but wouldn't the `TestPipelineRunName` always be removed [here](https://github.com/redhat-appstudio/integration-service/pull/408/files#diff-eeed1ae6f738d926a3e659cd3b10bbeb1f2c2a05f7ab22d899fa6914059660c0R122) before we check this?

Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go:133]: yes /o\

Comment (dirgim) [reply controllers/snapshot/snapshot_adapter.go:133]: That's a tough case actually. What if we don't allow rerunning Integration tests that are still running?

File: controllers/binding/snapshotenvironmentbinding_adapter.go [hunk 3]
```diff
@@ -75,11 +75,6 @@ func NewAdapter(snapshotEnvironmentBinding *applicationapiv1alpha1.SnapshotEnvir
 // EnsureIntegrationTestPipelineForScenarioExists is an operation that will ensure that the Integration test pipeline
 // associated with the Snapshot and the SnapshotEnvironmentBinding's IntegrationTestScenarios exist.
 func (a *Adapter) EnsureIntegrationTestPipelineForScenarioExists() (controller.OperationResult, error) {
-	if gitops.HaveAppStudioTestsFinished(a.snapshot) {
```
Comment (MartinBasti) [inline controllers/binding/snapshotenvironmentbinding_adapter.go:78]: This is my fix for the deadlock, I couldn't figure out what's the purpose of this gating.

Comment (dheerajodha) [reply controllers/binding/snapshotenvironmentbinding_adapter.go:78]: maybe because why would we want to create Integration PLRs for a snapshot that has already finished its testing?

Comment (MartinBasti) [reply controllers/binding/snapshotenvironmentbinding_adapter.go:78]: > maybe because why would we want to create Integration PLRs for a snapshot that has already finished its testing?

I'm re-running old tests in GH in order to figure out if it's regression, or race condition, or bug caused by something else, if new PRs are failing:D

File: controllers/snapshot/snapshot_adapter.go (hunk 2 above)
Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go]: Updated and unittest added

File: controllers/binding/snapshotenvironmentbinding_adapter_test.go [hunk 4]
```diff
@@ -410,33 +410,6 @@ var _ = Describe("Binding Adapter", Ordered, func() {
 
 	})
 
-	It("ensures the integrationTestPipelines are NOT created for a Snapshot that finished testing", func() {
```
Comment (MartinBasti) [inline controllers/binding/snapshotenvironmentbinding_adapter_test.go:413]: Surprisingly this test didn't fail yesterday

File: controllers/snapshot/snapshot_adapter.go [hunk 5]
```diff
@@ -75,6 +76,91 @@ func scenariosNamesToList(integrationTestScenarions *[]v1beta1.IntegrationTestSc
...
+
+	a.logger.Info("Re-running integration test for scenario", "scenario", scenarioName)
+
+	testStatuses, err := gitops.NewSnapshotIntegrationTestStatusesFromSnapshot(a.snapshot)
+	if err != nil {
+		return controller.RequeueWithError(err)
+	}
+
+	integrationTestScenarioStatus, ok := testStatuses.GetScenarioStatus(integrationTestScenario.Name)
+	if ok && integrationTestScenarioStatus.Status == intgteststat.IntegrationTestStatusInProgress {
```
Comment (dirgim) [inline controllers/snapshot/snapshot_adapter.go:109]: Thinking about this, maybe we also want to include the pending state here? 
If our re-reconciliation upon failing to remove the label starts before the PLR starts, we might find the status still in pending?

Comment (dirgim) [reply controllers/snapshot/snapshot_adapter.go:109]: To clarify, I'm thinking about the ephemeral environment scenario here, the static example will be in the `InProgress` state.

File: gitops/snapshot.go (hunk 1 above)
Comment (MartinBasti) [reply gitops/snapshot.go]: Replaced with operator-toolkit

File: controllers/snapshot/snapshot_adapter.go (hunk 5 above)
Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go]: That makes sense, added

File: controllers/snapshot/snapshot_adapter.go [hunk 6]
```diff
@@ -75,6 +76,92 @@ func scenariosNamesToList(integrationTestScenarions *[]v1beta1.IntegrationTestSc
 	return &result
 }
 
+// EnsureRerunPipelineRunsExist is responsible for recreating integration test pipelines triggered by users
+func (a *Adapter) EnsureRerunPipelineRunsExist() (controller.OperationResult, error) {
+
+	scenarioName, ok := gitops.GetIntegrationTestRunLabelValue(*a.snapshot)
+	if !ok {
+		// no test rerun trigerred
```
Comment (hongweiliu17) [inline controllers/snapshot/snapshot_adapter.go:84]:
```suggestion
		// no test rerun triggered
```

File: gitops/snapshot.go [hunk 7]
```diff
@@ -585,6 +588,26 @@ func HasSnapshotTestAnnotationChanged(objectOld, objectNew client.Object) bool {
 	return false
 }
 
+// HasSnapshotRerunLabelChanged returns a boolean indicating whether the Snapshot annotation for re-running
```
Comment (hongweiliu17) [inline gitops/snapshot.go:591]:
```suggestion
// HasSnapshotRerunLabelChanged returns a boolean indicating whether the Snapshot label for re-running
```

File: gitops/snapshot.go [hunk 8]
```diff
@@ -42,6 +42,9 @@ const (
 	// SnapshotTypeLabel contains the type of the Snapshot.
 	SnapshotTypeLabel = "test.appstudio.openshift.io/type"
 
+	// SnaphostIntegrationTestRun contains name of test we want to trigger run
+	SnaphostIntegrationTestRun = "test.appstudio.openshift.io/run"
```
Comment (hongweiliu17) [inline gitops/snapshot.go:46]:
```suggestion
	SnapshotIntegrationTestRun = "test.appstudio.openshift.io/run"
```

File: loader/loader.go [hunk 9]
```diff
@@ -443,3 +444,18 @@ func (l *loader) GetAutoReleasePlansForApplication(c client.Client, ctx context.
 
 	return &releasePlans.Items, nil
 }
+
+// GetScenario returns integration test scenario requested by name and namespace
+func (l *loader) GetScenario(c client.Client, ctx context.Context, name, namespace string) (*v1beta1.IntegrationTestScenario, error) {
+	scenario := &v1beta1.IntegrationTestScenario{}
```
Comment (hongweiliu17) [inline loader/loader.go:450]: We may replace these code in this function with the code like [this](https://github.com/redhat-appstudio/integration-service/blob/main/loader/loader.go#L118-L119) 
```
	scenario := &v1beta1.IntegrationTestScenario{}
	return scenario, toolkit.GetObject(name, namespace, c, ctx, scenario)
```

File: controllers/snapshot/snapshot_adapter.go [hunk 10]
```diff
@@ -75,6 +76,92 @@ func scenariosNamesToList(integrationTestScenarions *[]v1beta1.IntegrationTestSc
...
+
+	a.logger.Info("Re-running integration test for scenario", "scenario", scenarioName)
+
+	testStatuses, err := gitops.NewSnapshotIntegrationTestStatusesFromSnapshot(a.snapshot)
+	if err != nil {
+		return controller.RequeueWithError(err)
+	}
+
+	integrationTestScenarioStatus, ok := testStatuses.GetScenarioStatus(integrationTestScenario.Name)
+	if ok && (integrationTestScenarioStatus.Status == intgteststat.IntegrationTestStatusInProgress ||
```
Comment (hongweiliu17) [inline controllers/snapshot/snapshot_adapter.go:109]: This might be a corner case: Is there a user case that a user wants to rerun an inProgress PLR or trigger PLR for a pending integration test scenario that is not triggered yet due to the error in other integration test scenarios?

Comment (MartinBasti) [reply controllers/snapshot/snapshot_adapter.go:109]: We have a separate story to run integration tests independently, so one test wouldn't be blocked by error in the other.
I'd not solve this here but rather wait for the other story to be fixed.

Comment (hongweiliu17) [reply controllers/snapshot/snapshot_adapter.go:109]: I am fine with it as long as it is known.

**Comment by MartinBasti** [comment]
@hongweiliu17 PR updated with fixes, PTAL

**Review by hongweiliu17 — APPROVED** [review]
LGTM
